Make use of DeployToUpgrade

penev92 opened the pull request:
Enhance `DeployToUpgrade` and make use of it for the TS Mobile Sensor Array.

This is split into nice small commits that explain what is happening and why.
While reviewing by commits is not necessary here, it would likely be easier.

First three commits set the scene, next 4 commits add the necessary logic to `DeployToUpgrade`, and finally we have a use-case for finally using this trait.
(_[Awesomer Mobile Sensor Array](http://i.imgur.com/pBfXvvA.gif) is back!!_)

Notes:
- [x] The Mobile Sensor Array's deployed version (along with the deploy animation) has half a cell vertical offset. I noticed this after rebasing, and as shown in the gif, it wasn't an issue several months ago. Also not present in the normal `gadpsa`.
- [x] The original has `RadarInvisible=yes` as shown [here](https://github.com/LipkeGu/TSIniFiles/blob/4cc5adac88b012e27a6a82ba7c93d182451d9d64/rules.ini#L2073), so I've [removed](https://github.com/OpenRA/OpenRA/pull/9124/files#diff-20162b6ff85d8da38c40f5095744bad2R152) the `AppearsOnRadar` trait. But do we actually want this?

When this is all over I will remove the current mobile sensor array in favor of the new one (but keep the original name and tooltip, of course).
After that Tick tanks and Nod Artilleries can be done as well, thanks to @reaperrr's efforts (and mine :P ) toward making the necessary traits upgradable. (Ofc RA2 GIs can also be done now.)

Closes  #8932.

penev92 committed:
Add a deploy animation to DeployToUpgrade

Play a "deploy"/"undeploy" animation

penev92 committed:
Add facing to DeployToUpgrade

Make the actor turn to a desired facing before starting to deploy (and before granting the upgrade).

penev92 committed:
Don't get stuck in deployed mode because of an impossible move order

penev92 committed:
Make WithVoxelBody upgradable

penev92 committed:
Add a description to DisableUpgrade

penev92 committed:
Add deploy and undeploy sounds to DeployToUpgrade

Unit158 commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -32,6 +34,18 @@ public class DeployToUpgradeInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		[Desc("Cursor to display when unable to (un)deploy the actor.")]
 		public readonly string DeployBlockedCursor = "deploy-blocked";
 
+		[Desc("Facing that the actor must face before deploying. Set to -1 to not care about facing.")]
```
"not care" -> "ignore"?

Unit158 commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -32,6 +34,18 @@ public class DeployToUpgradeInfo : ITraitInfo, Requires<UpgradeManagerInfo>
+		public readonly int Facing = 0;
+
+		[Desc("Animation to play for deploying/undeploying.")]
+		public readonly string DeployAnimation = "";
```
Needs a sequence reference.

GraionDilach commented:
Animation and sounds should be optional imo.

GraionDilach commented on mods/ts/rules/shared-vehicles.yaml:
```diff
@@ -130,6 +130,59 @@ LPST:
+	DisableUpgrade:
+		UpgradeTypes: deployed
+		UpgradeMinEnabledLevel: 1
+		UpgradeMaxAcceptedLevel: 2
```
You shouldn't need these MaxAccepted: 2 statements - you're not even checking for them at all during revoking anyway.

GraionDilach commented:
The deployed sensor bug seems to be caused by the asset positioning differences between buildings and units, unit artworks are always centered on the actual positioon while buildings center their top cell's top point there You have to tweak the offset in yaml for that, I fear.

Compare the following two shots: 
![deploysensor](https://cloud.githubusercontent.com/assets/1136302/9414736/0ddd19e6-483b-11e5-8d64-3e16d0a6d25b.png) vs 
![wolverine](https://cloud.githubusercontent.com/assets/1136302/9414819/8e1e77f8-483b-11e5-874e-627cc09c0310.png)

Mailaender commented on mods/ts/rules/shared-vehicles.yaml:
```diff
@@ -130,6 +130,59 @@ LPST:
+		ROT: 5
+	RevealsShroud:
+		Range: 10c0
+	-AppearsOnRadar:	# TODO: Discuss whether we want to remove the trait. The original had RadarInvisible=yes
```
Let's spare the balancing discussion for now and just keep original Tiberian Sun behavior while we are still working on the technical parts.

penev92 replied:
I'll keep the comment for now, until I'm sure there will be no objections.

GraionDilach commented:
#8953 could be added in too - YR Siege Chopper comes into my mind.

penev92 commented:
The code here is still too young, and the aircraft code is undergoing a major overhaul (I have a PR that starts it, hopefully it will get finished eventually), so I'd like to skip this for now. Right now there is no non-hacky way of doing it.

GraionDilach commented:
Okay, I'm fine with that for now.

penev92 commented:
Made deploy animation and sounds optional (as they should have been) and removed `UpgradeMaxAcceptedLevel`.
Will keep looking into the offset, but will most likely need assistance with that.

GraionDilach commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -32,6 +34,18 @@ public class DeployToUpgradeInfo : ITraitInfo, Requires<UpgradeManagerInfo>
+		public readonly string DeploySound = "";
+
+		[Desc("Sound to play when undeploying.")]
+		public readonly string UnDeploySound = "";
```
Undeploy, imo.

GraionDilach commented:
Remove the defined Offsets in `gadpsa`'s sequence to fix that offset issue, just checked.

GraionDilach commented:
Copy the offset of `gadpsa:Defaults` to `lpst2` to fix the offset issue. I'm stupid. >_>

penev92 commented:
That fixed the offset for the actor, but screwed up the production icon :S

penev92 commented:
Removed the comment from `-AppearsOnRadar:` and renamed `UnDeploySound`.

penev92 commented:
Changed default values of `DeployAnimation`, `DeploySound`, `UndeploySound` and `Facing` to `null`.

LipkeGu commented:
@penev92 : sorry this was an bad asume that lpst is using place2 as deploy... not all units are using place2 as deploy sound. the lpst is using clicky other tanks are using place2

pchote commented on OpenRA.Game/Traits/BodyOrientation.cs:
```diff
@@ -69,7 +70,7 @@ public BodyOrientation(ActorInitializer init, BodyOrientationInfo info)
 					return info.QuantizedFacings;
 
-				var qboi = self.Info.Traits.GetOrDefault<IQuantizeBodyOrientationInfo>();
+				var qboi = self.Info.Traits.WithInterface<IQuantizeBodyOrientationInfo>().FirstOrDefault();
```
Is this still needed?

pchote replied:
Taking the first IQBOI is a bad idea, but IIRC @reaperrr has already worked around needing to do this for sprite code.

pchote replied:
See #9004.

pchote commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -32,6 +34,18 @@ public class DeployToUpgradeInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		public readonly string DeployBlockedCursor = "deploy-blocked";
 
+		[Desc("Facing that the actor must face before deploying. Will deploy disregarding facing by default.")]
+		public readonly int? Facing = null;
```
IIRC convention is to use `-1` as the magic value for automatic / no facing.

penev92 replied:
It was, but @Phrohdoh made a somewhat good case for using `int?` instead, and thus not needing to use (the otherwise valid value) -1 as a magic invalid default.

pchote replied:
IMO it would be better to keep this consistent with the rest of our code.  Somebody could argue for changing all of them in a different pr, but the other parts of the code may run into performance regressions due to unboxing.

GraionDilach replied:
I'm not certain if that can be disabled when inherited though.

penev92 replied:
Yeah that is a problem which I noticed when testing the suggestion. Will revert to -1  :/

pchote commented on OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs:
```diff
@@ -309,7 +309,7 @@ public void RefreshIcons()
 				var rect = new Rectangle(rb.X + x * (IconSize.X + IconMargin.X), rb.Y + y * (IconSize.Y + IconMargin.Y), IconSize.X, IconSize.Y);
 
-				var rsi = item.Traits.Get<RenderSpritesInfo>();
+				var rsi = item.Traits.WithInterface<RenderSpritesInfo>().First(t => t.IsTraitEnabled());
```
Actors should never have more than one `RenderSprites` trait, so this is not necessary.

penev92 commented:
I guess this depends on #9004 so I can remove the `BodyOrientation` change :/

reaperrr commented:
Needs rebase (#9004 was merged).

Phrohdoh commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -32,6 +34,18 @@ public class DeployToUpgradeInfo : ITraitInfo, Requires<UpgradeManagerInfo>
+		public readonly int? Facing = null;
+
+		[Desc("Animation to play for deploying/undeploying.")]
+		[SequenceReference] public readonly string DeployAnimation = null;
```
Nitpick: Could you combine these two attributes?

GraionDilach replied:
Why?

Phrohdoh replied:
So the `DeployAnimation` is on the same indentation level as the fields.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -75,11 +89,52 @@ public void ResolveOrder(Actor self, Order order)
+						return;
+					}
+
+					var render = self.TraitOrDefault<ISpriteBody>();
```
Why are you now restricting this to Sprites?

penev92 replied:
Nobody said this is the final version of the trait, but without a test case I can't reliably make it support every possible scenario. Also it would be much easier to digest if it's not all at once, I assume.

Phrohdoh replied:
I never implied it was the final version.
But this cannot then be used on most of our deployers (voxels), as far as I know.

penev92 replied:
All of them do Voxel -> Sprite -> Voxel (with Thumpers and GIs being the Sprite -> Sprite exception), so this will actually be good enough for the official mods (custom mods may want Voxel -> Voxel, with a custom MakeAnimation on the deployed voxel, but all of our cases will work.

GraionDilach replied:
Siege Chopper was voxel-voxel, but YR is out of the scope for now I guess.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs:
```diff
@@ -75,11 +89,52 @@ public void ResolveOrder(Actor self, Order order)
+				self.CancelActivity();
+
+				// Turn
+				if (info.Facing.HasValue && self.HasTrait<IFacing>())
```
This `HasTrait` can probably be cached.

penev92 commented:
- Reverted `Facing` to being `int`.
 - Beautified `DeployAnimation`'s attributes.
 - Removed the second commit "Don't crash on multiple `RenderSprites` traits" (an actor should not have multiple of those).
 - Removed the third commit "Don't crash on multiple traits implementing `IQuantizeBodyOrientationInfo`" (now that #9004 is merged there is no more need).
 - Switched from inheriting `^Vehicle` to `^VoxelVehicle`.
 - Added shadow to the deployed version.
 - Cached trait lookups for both `IFacing` and `ISpriteBody`.

Mailaender commented:
`LPST2` does not show a cloak reveal range circle. Seems to be otherwise fine.

GraionDilach commented:
The circle issue needs #8761 I guess.

penev92 commented:
It does display the reveal range. It's just a lot bigger than the one on `LPST` - 25 cells instead of 6. [Reference](https://github.com/LipkeGu/TSIniFiles/blob/4cc5adac88b012e27a6a82ba7c93d182451d9d64/rules.ini#L4252).
Also #8761 is not related to the `LPST`.  ;)

GraionDilach commented:
Wait, that's just 18 (17.6) cells in OpenRA terms!

Again, OpenRA measures the diagonals, while WW measured the sides...

penev92 commented:
I knew there was something in the range conversion I was missing, but there was nobody online to ask at the time...
Thanks, will fix shortly.

reaperrr commented:
The move order support is not a blocker, :+1:  once the yaml changes have been made.

penev92 committed:
Convert the Mobile Sensor Array to use DeployToUpgrade

Stop using Transforms.

penev92 commented:
Fixed the ranges (detection *and* sight) and removed the duplicate actor and sequences.
Haven't addressed the automatic undeploy comment as I'm currently not sure how to proceed with that.

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/afdcdb050cea0b41ab9f94c6117e41768ead181f...2efcf22959d4fa2bbec1de835aece1b6c4fbb94c)
